cocotb/types/array.py

```diff
@@ -0,0 +1,329 @@
+from sys import maxsize
+
+
+class Array(Sequence):
+    r"""
+    Fixed-size, arbitrarily-indexed, heterogeneous sequence type.
+
+    Arrays are similar to, but different from Python :class:`list`\ s.
+    An array can store values of any type or values of multiple types at a time, just like a :class:`list`.
+    Unlike :class:`list`\ s, an array's size cannot change.
+
+    The indexes of an array can start or end at any integer value, they are not limited to 0-based indexing.
+    Indexing schemes can be either ascending or descending in value.
+    An array's indexes are described using a :class:`~cocotb.types.Range` object.
+    Initial values are treated as iterables, which are copied into an internal buffer.
+
+    .. code-block:: python3
+
+        >>> Array("1234")  # the 0-based range `(0, len(value)-1)` is inferred
+        Array(['1', '2', '3', '4'], Range(0, 'to', 3))
+
+        >>> Array(range=Range(0, "downto", -3))  # the initial values are `None`
+        Array([None, None, None, None], Range(0, 'downto', -3))
+
+        >>> Array([1, True, None, "example"], Range(-2, 1))  # initial value and range lengths must be equal
+        Array([1, True, None, 'example'], Range(-2, 'to', 1))
+
+    Arrays also support "null" ranges; "null" arrays have zero length and cannot be indexed.
+
+    .. code-block:: python3
+
+        >>> Array(range=Range(1, "to", 0))
+        Array([], Range(1, 'to', 0))
+
+    Indexing and slicing is very similar to :class:`list`\ s, but it uses the indexing scheme specified.
+    Slicing, just like the :class:`~cocotb.types.Range` object uses an inclusive right bound,
+    which is commonly seen in HDLs.
+    Like :class:`list`\ s, if a start or stop index is not specified, it is inferred as the start or end of the array.
+    Slicing an array returns a new :class:`~cocotb.types.Array` object, whose bounds are the slice indexes.
+
+    .. code-block:: python3
+
+        >>> a = Array("1234abcd")
+        >>> a[7]
+        'd'
+        >>> a[2:5]
+        Array(['3', '4', 'a', 'b'], Range(2, 'to', 5))
+        >>> a[2:5] = reversed(a[2:5])
+        >>> "".join(a)
+        '12ba43cd'
+
+        >>> b = Array("1234", Range(0, -3))
+        >>> b[-2]
+        '3'
+        >>> b[-1:]
+        Array(['2', '3', '4'], Range(-1, 'downto', -3))
+        >>> b[:] = reversed(b)
+        >>> b
+        Array(['4', '3', '2', '1'], Range(0, 'downto', -3))
+
+    .. warning::
+        Arrays behave differently in certain situations than Python's builtin sequence types (:class:`list`, :class:`tuple`, etc.).
+
+        - Arrays are not necessarily 0-based and slices use inclusive right bounds,
+          so many functions that work on Python sequences by index (like :mod:`bisect`) may not work on arrays.
+        - Slice indexes must be specified in the same direction as the array and do not support specifying a "step".
+        - When setting a slice, the new value must be an iterable of the same size as the slice.
```

Comment:
```suggestion
        - When setting a slice, the new value must be an Array of the same size as the slice.
```

Reply:
Is this something you want to enforce?

Reply:
At first I thought it was inconsistent because `Logic` can only be combined with `Logic`, but slice assignment on `Array` can work with non-`Array`s? We also decided to not support implicit conversion to `Logic` in binary operations with `Logic`, but for `LogicArray` implicit casting is either required, or we only support assigning with other `LogicArray`s, which is just totally inconsistent.

Reply:
Basically it comes down to whether we want to support implicit casting of `Iterable[Any]` in `LogicArray` slice assignments or not. I'm choosing the "no implicit casting" solution because that is what we did for `Logic`.

Reply:
`Iterable[Any]` is fine because `Array`s are that anyways.